src/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -71,12 +71,22 @@ public unsafe static String GetEnvironmentVariable(String variable)
             return Encoding.UTF8.GetString((byte*)result, size);
         }
 
-        public static string MachineName
+        private const int MAX_HOST_NAME = 256; // 255 max and null 
+        public static unsafe string MachineName
         {
             get
             {
-                // UNIXTODO: Not yet implemented. Issue: dotnet/corert#650.
-                throw new NotImplementedException();
+                byte *hostName = stackalloc byte[MAX_HOST_NAME];
+                int hostNameLength = Interop.Sys.GetMachineName(hostName, MAX_HOST_NAME);
+                if (hostNameLength < 0)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_ComputerName);
+                }
+
+                if (hostNameLength == 0)
```

Comment:
I looked at the implementation of GetString and it go deep in the stack before checking for zero length so I decided to do the check here better